sparse/client.go

```diff
@@ -51,207 +47,184 @@ func syncFile(localPath string, addr TCPEndPoint, remotePath string, timeout int
 				}
 			}
 		}
 	}
 
 	// special case, the whole file is a hole
 	if len(exts) == 0 && fileSize != 0 {
-		// report hole
 		holeInterval = Interval{0, fileSize}
-		log.Debugf("The file is a hole: %s", holeInterval)
+		log.Infof("The file is a hole: %s", holeInterval)
 
 		// syncing hole interval
-		err := session.SyncHoleInterval(holeInterval)
+		err := client.syncHoleInterval(holeInterval)
 		if err != nil {
-			log.Debugf("SyncHoleInterval failed: %s", holeInterval)
-			return err
+			return fmt.Errorf("syncHoleInterval %s failed, err: %s", holeInterval, err)
 		}
 	}
 
-	return session.send(requestHeader{requestMagic, syncDone})
+	return nil
 }
 
-func (session *SyncSessionClient) sendSyncRequest(path string, size int64) error {
-	err := session.send(requestHeader{requestMagic, syncRequestCode})
-	if err != nil {
-		log.Debug("send requestHeader failed: ", err)
-		return err
-	}
-	err = session.send(path)
-	if err != nil {
-		log.Debug("send path failed: ", err)
-		return err
+func (client *syncClient) sendHTTPRequest(method string, action string, interval Interval, data []byte) (*http.Response, error) {
+	httpClient := &http.Client{Timeout: time.Duration(httpClientTimeout * time.Second)}
+
+	url := fmt.Sprintf("http://%s/v1-ssync/%s", client.remote, action)
+
+	var req *http.Request
+	if data != nil {
+		req, _ = http.NewRequest(method, url, bytes.NewBuffer(data))
+	} else {
+		req, _ = http.NewRequest(method, url, nil)
 	}
-	err = session.send(size)
-	if err != nil {
-		log.Debug("send size failed: ", err)
-		return err
+
+	req.Header.Add("Accept", "application/json")
+
+	q := req.URL.Query()
+	q.Add("begin", strconv.FormatInt(interval.Begin, 10))
+	q.Add("end", strconv.FormatInt(interval.End, 10))
+	req.URL.RawQuery = q.Encode()
+
+	log.Debugf("method: %s, url with query string: %s, data len: %d", method, req.URL.String(), len(data))
+
+	return httpClient.Do(req)
+}
+
+func (client *syncClient) openServer() error {
+	var err error
+	var resp *http.Response
+
+	timeStart := time.Now()
+	timeStop := timeStart.Add(time.Duration(client.timeout) * time.Second)
+	for timeNow := timeStart; timeNow.Before(timeStop); timeNow = time.Now() {
+		resp, err = client.sendHTTPRequest("GET", "open", Interval{0, client.fileSize}, nil)
+		if err == nil {
+			break
+		}
+		log.Warnf("Failed to open server: %s, Retrying...", client.remote)
+		if timeNow != timeStart {
+			// only sleep after the second attempt to speedup tests
+			time.Sleep(1 * time.Second)
+		}
 	}
 
-	var ack bool
-	err = session.receive(&ack)
 	if err != nil {
-		log.Debug("decode ack from server failed: ", err)
-		return err
+		return fmt.Errorf("open failed, err: %s", err)
 	}
-	log.Debugf("got SyncRequest ack back: %v", ack)
-	if ack == false {
-		return fmt.Errorf("Got negative ack from server")
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("resp.StatusCode(%d) != http.StatusOK", resp.StatusCode)
 	}
+	defer resp.Body.Close()
```

Comment:
Wrong place to defer. You don't put `defer` in the last line of function, but in when you got the resource.